src/libbson/src/bson/bson-error.c

```diff
@@ -98,25 +103,47 @@ bson_set_error (bson_error_t *error, /* OUT */
  */
 
 char *
-bson_strerror_r (int err_code,  /* IN */
-                 char *buf,     /* IN */
-                 size_t buflen) /* IN */
+bson_strerror_r (int err_code,                    /* IN */
+                 char *buf BSON_MAYBE_UNUSED,     /* IN */
+                 size_t buflen BSON_MAYBE_UNUSED) /* IN */
 {
    static const char *unknown_msg = "Unknown error";
    char *ret = NULL;
 
 #if defined(_WIN32)
+   // Windows does not provide `strerror_l` or `strerror_r`, but it does
+   // unconditionally provide `strerror_s`.
    if (strerror_s (buf, buflen, err_code) != 0) {
       ret = buf;
    }
-#elif defined(BSON_HAVE_XSI_STRERROR_R)
-   if (strerror_r (err_code, buf, buflen) == 0) {
-      ret = buf;
+#elif defined(__APPLE__)
+   // Apple does not provide `strerror_l`, but it does unconditionally provide
+   // the XSI-compliant `strerror_r`, but only when compiling with Apple Clang.
+   // GNU extensions may still be a problem if we are being compiled with GCC on
+   // Apple. Avoid the compatibility headaches with GNU extensions and the musl
+   // library by assuming QoI will not cause UB when reading the error message
+   // string even when `strerror_r` fails.
+   (void) strerror_r (err_code, buf, buflen);
+#elif defined(_XOPEN_SOURCE) && _XOPEN_SOURCE >= 700
+   // Avoid `strerror_r` compatibility headaches with GNU extensions and the
+   // musl library by using `strerror_l` instead. Furthermore, `strerror_r` is
+   // scheduled to be marked as obsolete in favor of `strerror_l` in the
+   // upcoming POSIX Issue 8 (see:
+   // https://www.austingroupbugs.net/view.php?id=655).
+   //
+   // POSIX Spec: since strerror_l() is required to return a string for some
+   // errors, an application wishing to check for all error situations should
+   // set errno to 0, then call strerror_l(), then check errno.
+   errno = 0;
+   ret = strerror_l (err_code, uselocale ((locale_t) 0));
```

Comment:
[`strerror(3)`](https://man7.org/linux/man-pages/man3/strerror.3.html) states:

> The behavior of strerror_l() is undefined if locale is the special locale object LC_GLOBAL_LOCALE or is not a valid locale object handle.

Quoting [`uselocale(3)`](https://man7.org/linux/man-pages/man3/uselocale.3.html):

> On success, uselocale() returns the locale handle that was set by the previous call to uselocale() in this thread, or LC_GLOBAL_LOCALE if there was no such previous call. On error, it returns (locale_t) 0, and sets [errno](https://man7.org/linux/man-pages/man3/errno.3.html) to indicate the error.

I don't see any other calls to `uselocale()` throughout libbson and libmongoc. Do we have to worry about `LC_GLOBAL_LOCALE` being returned? If the intention is to disregard a locale altogether, then would it make more sense to just pass `(locale_t) 0` directly to `strerror_l` instead?

I don't have much experience with `uselocale()`, but a related discussion that came up while searching for this `LC_GLOBAL_LOCALE` edge case was: https://austin-group-l.opengroup.narkive.com/MFGftIBP/use-of-lc-global-locale-with-l-functions

Reply:
Good call. Implemented the missing locale handling to ensure we always pass a valid locale.